akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -961,18 +965,115 @@ abstract class GraphStageLogic private[stream] (val inCount: Int, val outCount:
     }
   }
 
+  private class AsyncCallbackAsMaterializedValueImpl[T](handler: T ⇒ Unit) extends AsyncCallbackAsMaterializedValue[T] {
+    trait CallbackState
+    case class NotInitialized(list: List[T]) extends CallbackState
+    case object Initialized extends CallbackState
+    case object Stopped extends CallbackState
+
+    val concurrentSet = ConcurrentHashMap.newKeySet[Promise[_]]()
+
+    private[this] val callbackState = new AtomicReference[CallbackState](NotInitialized(Nil))
```

Comment:
looks like `callbackState` is only accessed within `synchronized` so then there is no need for `AtomicReference`